### DISCUSSION TITLE
Upper scope

### DISCUSSION THREAD

**Discussion by ghost** [h3rald/min#117] [Features: Implemented]
How could i access *upper scope* from a *current scope*, in min?

**Comment by h3rald** [comment]
Alright... so, the `scope` symbol pushes a dictionary with a reference of the current scope on the stack, so:

    27 :a scope :s (2 *s/a +) ->  
    ; 29

Also, you can use `ROOT` to access the root scope.

**Reply by ghost** [reply]
The important point is accessing upper scope. So `ROOT` is not a proper solution for this. `scope :s` is more good but not perfect way to do it. So i think there should be a native symbol to provide upper accessing.

**Reply by ghost** [reply]
Because of scope lookout symbols.

**Reply by h3rald** [reply]
Alright that could be implemented but... what would be the main use case for it? The problem is that I can get the upper scope but sometimes you may want a scope at a higher level as well so... that's why with `scope` you can set a reference to whatever scope you want to capture and then use it later on.

What would be the benefit of exposing `parent-scope` for example?

**Reply by ghost** [reply]
Every scope should have own mutable names, after that we would understand where we are.

ROOT -> Scope1 -> Scope2 -> Scope3 -> ...

Root means as the main file. Other scopes may acts like subfiles?

That may facilitate some min file processes.(f reading... f writing... f executing... f evaluating...  symbol lifting...)

**Reply by h3rald** [reply]
Well, a new scope is created for *every quotation* and of course loaded/required files. Actually when I first implemented lexical scoping scopes *did* have a unique, auto-generated name. But then I didn't have any practical use for them and also I didn't like the idea of adding so many strings to memory.

I dunno about scope navigation. Maybe some way to go up the scope chain like a `parent-scope` symbol (which would take a scope as input and return its parent as output, so you could chain it), but what would be the practical use for it?

Also about storing extra pointless strings: right now to ensure proper error reporting I am storing the file name, row, and column of *every symbol instance* right after parsing. I wish there could be a better way to manage this but I am not sure how to manage it. 😞

**Reply by ghost** [reply]
> Well, a new scope is created for every quotation and of course loaded/required files.

Should be no new scope creating for files as my new idea. If every scope has names, there will be scope names of files too. That means `ROOT` would behave like a cheap usement of file scopes. So all shell configuration files would drop to `.min` files. 

**Achievement gathered: Easy-Peasy Configuration/Linking/Min File Interactions Handling** 😄

**Reply by h3rald** [reply]
OK, no, I don't understand 😳

Please elaborate: what are you suggesting? At present a new scope is created when:
* dequoting/applying/mapping etc.
* requiring/loading new files
* ...

So what you are saying is already happening. I tried adding names but... there's simply too many scopes to be able to reason about; some are "more important" than others like the scope created for the module resulting from requiring a file... now that's probably worth exploring, but I am not sure on what you mean exactly. 🤔

On another note, v0.31.0 is out, and that's probably going to hurt a bit 🤕 -- but hopefully after this there shouldn't be many more breaking changes 😇

**Reply by ghost** [reply]
Yes, you're right! A very breaky changes comes up with the new version in min. But no problem. You're doing so amazingly!! 
👍 👍 

> requiring/loading new files

I don't like current module implementing system. It is dirty. **Need more Doc**
But i didn't try v0.31.0 yet. ⚓

> but I am not sure on what you mean exactly

I'm thinking how.. , i can explain what i thought:
1. Eliminating the differencies of files and scopes
2. Using better to handle with modules and getting rid of `require`, `load`
3. Calling modules with their scope names to use them in files
4. ...others...

**Reply by h3rald** [reply]
> 
> I don't like current module implementing system. It is dirty. **Need more Doc**
> 
> But i didn't try v0.31.0 yet. ⚓

I removed `module` in 0.31.0: you can create modules automatically by using `require` to load a file containing min symbol definitions. Much easier and tidier I think. Plus you get stack pollution checks.

> 
> 
> > but I am not sure on what you mean exactly
> 
> 
> 
> I'm thinking how.. , i can explain what i thought:
> 
> 1. Eliminating the differencies of files and scopes
> 
Stop right there: there has to be a new scope created for every quotation, otherwise things will be counter intuitive. What do you mean by this?


> 2. Using better to handle with modules and getting rid of `require`, `load`
> 
What's wrong with `require`? What would you use instead?

> 3. Calling modules with their scope names to use them in files
> 
Ok, I think I start to understand maybe! Are you thinking of something similar to the invoke syntax, but doing lookups on files instead of dictionaries??

> 4. ...others...

Please write an example for all this! That would be very helpful. This does indeed sound interesting! 🤔

**Reply by ghost** [reply]
> what would be the main use case for it?
> Please write an example for all this! That would be very helpful. This does indeed sound interesting! 🤔

Look at #125

**Reply by ghost** [reply]
Upper scope accessment would be also useful to lift lambdas or control lambdas or symbols etc. ...
See here -> #123
